x/interchainquery/keeper/msg_server.go

```diff
@@ -28,105 +28,143 @@ func NewMsgServerImpl(keeper Keeper) types.MsgServer {
+		if !found {
+			return fmt.Errorf("unable to fetch client state")
 		}
+		path := commitmenttypes.NewMerklePath([]string{pathParts[1], url.PathEscape(string(q.Request))}...)
 
-		// PROCESS QUERY RESPONSE
-		pathParts := strings.Split(q.QueryType, "/")
+		k.Logger(ctx).Info(fmt.Sprintf("[MOOSE] proof is %#v", msg.ProofOps))
+		merkleProof, err := commitmenttypes.ConvertProofs(msg.ProofOps)
+		if err != nil {
+			return fmt.Errorf("error converting proofs")
+		}
 
-		// verify the query results are valid by checking the assocaited proof
-		if pathParts[len(pathParts)-1] == "key" {
-			if msg.ProofOps == nil {
-				return nil, fmt.Errorf("unable to validate proof. No proof submitted")
-			}
-			connection, _ := k.IBCKeeper.ConnectionKeeper.GetConnection(ctx, q.ConnectionId)
+		tmclientstate, ok := clientState.(*tmclienttypes.ClientState)
+		if !ok {
+			return fmt.Errorf("error unmarshaling client state %v", clientState)
+		}
 
-			msgHeight, err := cast.ToUint64E(msg.Height)
-			if err != nil {
-				return nil, err
+		if len(msg.Result) != 0 {
+			// if we got a non-nil response, verify inclusion proof.
+			if err := merkleProof.VerifyMembership(tmclientstate.ProofSpecs, consensusState.GetRoot(), path, msg.Result); err != nil {
+				return fmt.Errorf("unable to verify proof: %s", err)
 			}
-			height := clienttypes.NewHeight(clienttypes.ParseChainID(q.ChainId), msgHeight+1)
-			consensusState, found := k.IBCKeeper.ClientKeeper.GetClientConsensusState(ctx, connection.ClientId, height)
+			k.Logger(ctx).Info(fmt.Sprintf("Proof validated! module: %s, queryId %s", types.ModuleName, q.Id))
 
-			if !found {
-				return nil, fmt.Errorf("unable to fetch consensus state")
+		} else {
+			// if we got a nil response, verify non inclusion proof.
+			if err := merkleProof.VerifyNonMembership(tmclientstate.ProofSpecs, consensusState.GetRoot(), path); err != nil {
+				return fmt.Errorf("unable to verify proof: %s", err)
 			}
+			k.Logger(ctx).Info(fmt.Sprintf("Non-inclusion Proof validated! module: %s, queryId %s", types.ModuleName, q.Id))
+		}
+	}
+	return nil
+}
 
-			clientState, found := k.IBCKeeper.ClientKeeper.GetClientState(ctx, connection.ClientId)
-			if !found {
-				return nil, fmt.Errorf("unable to fetch client state")
-			}
-			path := commitmenttypes.NewMerklePath([]string{pathParts[1], url.PathEscape(string(q.Request))}...)
+// call the query's associated callback function
+func (k Keeper) FindAndInvokeCallback(ctx sdk.Context, msg *types.MsgSubmitQueryResponse, q types.Query) error {
+	// get all the stored queries and sort them for determinism
+	moduleNames := []string{}
+	for moduleName := range k.callbacks {
+		moduleNames = append(moduleNames, moduleName)
+	}
+	sort.Strings(moduleNames)
+
+	for _, module := range moduleNames {
+		k.Logger(ctx).Info(fmt.Sprintf("Executing callback for queryId (%s), module (%s)", q.Id, module))
+		moduleCallbackHandler := k.callbacks[module]
```

Comment:
@asalzmann @riley-stride We briefly discussed a way to refactor these ICQs the other day. Definitely does not have to be done in this PR, but can someone jog my memory so we don't lose track of it! I thought it was something related to not having to loop through each module?

Reply:
I think we could store the calling module on the ICQ! Ideally though, you'd need to also verify that the ICQ originated from the calling module - which is a bit of a harder problem